CMakeLists.txt

```diff
@@ -0,0 +1,294 @@
+#       disclaimer   in  the  documentation  and/or  other  materials
+#       provided with the distribution.
+#
+#     * Neither  the name of the <ORGANIZATION> nor the names of  its
+#       contributors  may  be  used to endorse  or  promote  products
+#       derived  from  this software without specific  prior  written
+#       permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+# "AS  IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT  NOT
+# LIMITED  TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND  FITNESS
+# FOR  A  PARTICULAR  PURPOSE ARE DISCLAIMED. IN NO EVENT  SHALL  THE
+# COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+# INCIDENTAL,    SPECIAL,   EXEMPLARY,   OR   CONSEQUENTIAL   DAMAGES
+# (INCLUDING,  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+# SERVICES;  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+# HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+# STRICT  LIABILITY,  OR  TORT (INCLUDING  NEGLIGENCE  OR  OTHERWISE)
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED
+# OF THE POSSIBILITY OF SUCH DAMAGE.
+#
+cmake_minimum_required(VERSION 3.3)
+
+project(uriparser
+    VERSION
+        0.9.1
+)
+
+# See https://www.gnu.org/software/libtool/manual/html_node/Updating-version-info.html
+set(URIPARSER_SO_CURRENT    1)
+set(URIPARSER_SO_REVISION  24)
+set(URIPARSER_SO_AGE        0)
+
+include(CheckFunctionExists)
+include(CheckSymbolExists)
+include(GNUInstallDirs)
+
+#
+# Configuration
+#
+option(BUILD_SHARED_LIBS "Build shared libraries (rather than static ones)" ON)
+option(URIPARSER_BUILD_DOCS "Build API documentation (requires Doxygen, Graphviz, and (optional) Qt's qhelpgenerator)" ON)
+option(URIPARSER_BUILD_TESTS "Build test suite (requires GTest >=1.8.1)" ON)
+option(URIPARSER_BUILD_CHAR "Build code supporting data type 'char'" ON)
+option(URIPARSER_BUILD_WCHAR_T "Build code supporting data type 'wchar_t'" ON)
+
+if(NOT URIPARSER_BUILD_CHAR AND NOT URIPARSER_BUILD_WCHAR_T)
+    message(FATAL_ERROR "One or more of URIPARSER_BUILD_CHAR and URIPARSER_BUILD_WCHAR_T needs to be enabled.")
+endif(NOT URIPARSER_BUILD_CHAR AND NOT URIPARSER_BUILD_WCHAR_T)
+if(URIPARSER_BUILD_TESTS AND NOT (URIPARSER_BUILD_CHAR AND URIPARSER_BUILD_WCHAR_T))
+    message(FATAL_ERROR "URIPARSER_BUILD_TESTS=ON requires both URIPARSER_BUILD_CHAR=ON and URIPARSER_BUILD_WCHAR_T=ON.")
+endif(URIPARSER_BUILD_TESTS AND NOT (URIPARSER_BUILD_CHAR AND URIPARSER_BUILD_WCHAR_T))
+
+if(NOT URIPARSER_BUILD_CHAR)
+    # NOTE that add_compile_definitions would require CMake >=3.12
+    add_definitions(-DURI_NO_ANSI)
+endif(NOT URIPARSER_BUILD_CHAR)
+if(NOT URIPARSER_BUILD_WCHAR_T)
+    # NOTE that add_compile_definitions would require CMake >=3.12
+    add_definitions(-DURI_NO_UNICODE)
+endif(NOT URIPARSER_BUILD_WCHAR_T)
+
+#
+# config.h
+#
+# NOTE that add_compile_definitions would require CMake >=3.12
+add_definitions(-DHAVE_CONFIG_H)
```

Comment:
You can use `target_compile_definitions` on the target instead.

Reply:
I have dropped use of `HAVE_CONFIG_H` altogether just now.
For the other two cases I have moved to `target_compile_definitions` as suggest, and agree about the "global is not cool" part.

I'm happy to move `-DURI_NO_ANSI` into `config.h` at some point, but maybe not before this merge to limit the scope and not touch everything at once.
By toolchain file you seem to refer to something other than `config.h`?